src/JustSaying/Messaging/Middleware/Handle/HandlerInvocationMiddleware.cs

```diff
@@ -0,0 +1,44 @@
+using System;
+using System.Reflection;
+using System.Threading;
+using System.Threading.Tasks;
+using JustSaying.Messaging.MessageHandling;
+
+namespace JustSaying.Messaging.Middleware.Handle
+{
+    public class HandlerInvocationMiddleware<T> : MiddlewareBase<HandleMessageContext, bool>
+    {
+        private readonly Func<HandlerResolutionContext, IHandlerAsync<T>> _handlerResolver;
+
+        public HandlerInvocationMiddleware(Func<HandlerResolutionContext, IHandlerAsync<T>> handlerResolver)
+        {
+            _handlerResolver = handlerResolver;
+        }
+
+        protected override async Task<bool> RunInnerAsync(
+            HandleMessageContext context,
+            Func<CancellationToken, Task<bool>> func,
+            CancellationToken stoppingToken)
+        {
+            if (context == null) throw new ArgumentNullException(nameof(context));
+
+            stoppingToken.ThrowIfCancellationRequested();
+
+            var resolutionContext = new HandlerResolutionContext(context.QueueName);
+
+            IHandlerAsync<T> handler = _handlerResolver(resolutionContext);
+
+            //TODO: cache these
+            Type handlerType = handler.GetType();
```

Comment:
Can we instead use `var task = handler.Handle(context.MessageAs<T>());` and flow the generic constraint?

Reply:
Ahh, yes we can! I did it that way initially but then wanted to remove the generic parameter from `HandlerInvocationMiddleware<T>` for some reason, but that turned out to not be necessary.

Great stuff!

Reply:
Have fixed this up 👍